src/v/utils/tests/bottomless_token_bucket_test.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "utils/bottomless_token_bucket.h"
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_CASE(muldiv_test) {
+    constexpr static auto mn = std::numeric_limits<int64_t>::min();
+    constexpr static auto mx = std::numeric_limits<int64_t>::max();
+    // straight samples
+    BOOST_CHECK_EQUAL(muldiv(98733, 9287, 6489), 98733 * 9287 / 6489);
+    BOOST_CHECK_EQUAL(muldiv(-494162, -654, -998), -494162 * -654 / -998);
+    // zeros
+    BOOST_CHECK_EQUAL(muldiv(0, -654, -998), 0);
+    BOOST_CHECK_EQUAL(muldiv(-494162, 0, -998), 0);
+    // near direct overflow
+    BOOST_CHECK_EQUAL(muldiv(6935552231102021794, 2, 3), 4623701487401347862);
+    BOOST_CHECK_EQUAL(
+      muldiv(10, -2305878194122661888, -3), 7686260647075539626);
+    // near intermediate overflow
+    BOOST_CHECK_EQUAL(
+      muldiv(1152921504606846976, 8495162354, 1085720514), 9020972926972920768);
+    // boundary values as 1st argument
+    BOOST_CHECK_EQUAL(muldiv(mx, 1, 1), mx);
+    BOOST_CHECK_EQUAL(muldiv(mn, 1, 1), mn);
+    BOOST_CHECK_EQUAL(muldiv(mx, 777, 1554), mx / 2);
+    BOOST_CHECK_EQUAL(muldiv(mn, -32767, 32768), (1ull << 63) - (1ull << 48));
+    // boundary values as 2nd argument
+    BOOST_CHECK_EQUAL(muldiv(1, mx, 1), mx);
+    BOOST_CHECK_EQUAL(muldiv(1, mn, 1), mn);
+    // boundary values as 3rd argument
+    BOOST_CHECK_EQUAL(muldiv(1, 1, mx), 0);
+    BOOST_CHECK_EQUAL(muldiv(1, 1, mn), 0);
+    BOOST_CHECK_EQUAL(muldiv(mx / 2, 2, mx), 0); // mx is odd, mx/2 is truncated
+    BOOST_CHECK_EQUAL(muldiv(mn / 2, 2, mn), 1); // mn is even
+}
+
+BOOST_AUTO_TEST_CASE(test_bottomless_token_bucket) {
+    using namespace std::chrono;
+
+    // === straight cases
+    bottomless_token_bucket b1(1, milliseconds(1)), b2(1000, seconds(1));
+    BOOST_CHECK_EQUAL(b1.quota(), 1);
+    BOOST_CHECK_EQUAL(b1.tokens(), 0);
+    BOOST_CHECK_EQUAL(b2.quota(), 1000);
+    BOOST_CHECK_EQUAL(b2.tokens(), 0);
+
+    ss::lowres_clock::time_point now;
```

Comment:
wonder if you could initialize this with an explicit zero some how since that seems to be the intention?

Reply:
`time_point` [default ctor](https://en.cppreference.com/w/cpp/chrono/time_point/time_point) initializes it to zero.